### PR TITLE
Support TAP devices in "snabb-lwaftr run-nohw"

### PR DESCRIPTION
**Pull request by aperezdc**
Either side of the AFTR may use either a `RawSocket` (by prefixing the device
name with `raw:`) or attach to an existing TUN/TAP device using the `Tap` application (by prefixing the device name with `tap:`). Note that the TAP devices have to be configured manually e.g. using the `ip` command.

This is the same as PR #236, rebased on top of the `lwaftr` branch and squashed in a single commit.

### CODE REVIEW

File: src/program/lwaftr/run_nohw/run_nohw.lua [hunk 1]
```diff
@@ -31,63 +37,81 @@ local function parse_args(args)
...
          check(arg, "argument to '--inet-if' not specified")
-         inet_if = arg
+         inet_if_kind, inet_if = arg:match("^([a-z]+):([^%s]+)$")
+         check(inet_if,
+               "invalid/missing device name in '%s'", arg)
+         check(inet_if_kind and device_kind_map[inet_if_kind],
+               "invalid/missing device kind in '%s'", arg)
+         inet_if_kind = device_kind_map[inet_if_kind]
+      end;
+      D = function ()
```
Comment (dpino) [inline src/program/lwaftr/run_nohw/run_nohw.lua]: 'D' flag is used for duration in many of the lwaftr commands and other snabb commands. Is it possible to use other name? In addition to it, I see the what this flag does is to set `conf.debug`, so the same can be achieved through a lwAFTR.conf file that has `debug` set to true. If that's the case, I'd prefer to remove this flag as a parameter option.

Comment (aperezdc) [inline src/program/lwaftr/run_nohw/run_nohw.lua]: Sure thing, I will upload an additional commit to remove the `D` flag.

**Comment by aperezdc** [comment]
@dpino: Done, `run-nohw` no longer accepts a `-D` command line argument.

**Comment by wingo** [comment]
ignorant question :) why is run-nohw necessary?  would it be possible to make `run` detect when it is running on tap interfaces using the if_kind parsing that you do in this patch?

**Comment by aperezdc** [comment]
@wingo: Well, there's nothing preventing `snabb-lwaftr run` to handle TAP interfaces and raw sockets importing parts of the code from this patch. Should I go ahead and do just that? If yes, it would be funny to write `--v4-pci tap:devname`, so what would be do with the existing command line flags? The nicest thing would be to rename them as `--v4-dev <kind>:<options>` (and `--v6-dev <kind>:<options>`), but I don't know if we want to break the existing syntax. Another option is adding the new flags, and allowing the old ones as shorthands (so for example `--v4-pci <pci-addr>` would be interpreted as `--v4-dev intel82599:<pci-addr>`).

**Comment by wingo** [comment]
Hmm!  It's andy opinion time!  OK so I agree that --v4-pci is a bad option name.  Please rename to --v4-dev or just --v4, whichever as you like, updating documentation, and adding --v4-pci as an alias.  I think you can do that just by adding an entry to the long option table.  No need to warn for now; we need to make a new release and shouldn't introduce a warning now.  Regarding "kind"; I think that if the addr is a valid PCI address, then we should treat it as a HW card for now.  If that's possible that would be best.  We need to reduce the proliferation of similar programs in the lwaftr, it seems to me.

**Comment by aperezdc** [comment]
@wingo: I agree to all you wrote :wink:

If the argument to `--v4-dev` matches a pattern for a PCI device address, I will make the code assume that the `<kind>:` part is omitted, and implicitly select the Intel 82599 application.

**Comment by aperezdc** [comment]
PTAL. Now the functionality is merged into `snabb lwaftr run`.

Probably we want #323 merged beforehand, and rebasing this PR on top of it before merging. WDYT?

**Comment by dpino** [comment]
@aperezdc OK, but before merging I'm going to reintroduce the old arguments "--v4-pci" and "--v6-pci" for backwards compatibility.

**Comment by aperezdc** [comment]
Rebased and updated :ok_hand:

File: src/program/lwaftr/run/README [hunk 2]
```diff
@@ -1,15 +1,41 @@
...
+       -v                         Verbose (repeat for more verbosity).
+
+Deprecated arguments:
+       --v4-pci <pci-addr>        PCI device number for the INET-side NIC.
+       --v6-pci <pci-addr>        PCI device number for the B4-side NIC.
+       --virtio                   Use virtio-net interfaces instead of Intel 82599.
+
+Device kinds and configuration parameters:
+
+       intel82259:<pci-addr>
```
Comment (wingo) [inline src/program/lwaftr/run/README:25]: 82599, not 82259, here and below.

With PCI addresses, is it necessary to specify the kind?  AFAIU we should be able to get this information from /sys/.  See first snippet in https://github.com/lukego/blog/issues/13.

**Comment by dpino** [comment]
PR needs heavy rebase and testing. Closing it after discussion with @aperezdc on chat.

**Pull request closed without merging**